### DISCUSSION TITLE
Accessing batchools logs

### DISCUSSION THREAD

**Discussion by multimeric** [ropensci/targets#939] [Help]
I'm running using the batchools backend. However I can't find a high level way to input a target name and for `targets` to give me the batchtools logs. They are, however, definitely being stored, because I was able to find the logs for the target `loaded_count_db_f115fc0f` in the file `.future/20220928_123242-iHrqXC/loaded_count_db_f115fc0f_237605795/logs/job863ff95703eab508b4c60603c77b149e.log`.  It's a bit of a pain having to switch to a bash terminal and cat these files whenever I have an error, though.

**Comment by wlandau** [comment]
`targets` works exclusively through the officially supported `future` R package interface, and there are no plans specific `targets` functions to access particular logs of files generated by `batchtools`. You may be able to set flags in the template file to specify where logs from the cluster are stored. This will depend on the resource manager you have, but in SGE, the easiest way is to set the `-o` flag to something other than `/dev/null`.

https://github.com/ropensci/targets/blob/17a12db567c1e2b44e7d21c85e82dbf1cecfb0ba/inst/templates/clustermq/sge.tmpl#L5

**Reply by multimeric** [reply]
That's a shame, I wonder if interfacing directly with `batchtools` (rather than through `future`) would yield a tighter and more featurful integration. If I set the log file path in the template (which SLURM can do), and then pass in a path in the resources list, how can I then look up that path for an arbitrary job at a future date?

**Reply by wlandau** [reply]
I have no plans to integrate directly with `batchtools` or implement custom high-level functions to manage the log files generated by other packages. Such a feature would be out of scope. `targets` operates at a high level and only interacts with `future` through its officially supported interface. (Otherwise, `targets` would break if the internals of another package change. It's the same reason why it is not encouraged to call unexported package functions (using `:::`).) Any R functions for log management would need to be available through `future`, `batchtools`, or `future.batchtools` because those are the packages that are closer to the level at which those logs are generated.

**Reply by stuvet** [reply]
Can you think of any way to predict/retrieve the backend job id or name from within the `targets` workflow? Seems like this would allow log retrieval via a system call.

**Reply by wlandau** [reply]
tar_name() returns the name of the target currently running.

**Reply by stuvet** [reply]
Never mind. Submitted job names are the same as the targets, so to get only the jobid:

```
#!/bin/env bash

JOBNAME=mytarget_0888390

sacct -n -X --format jobid --name $JOBNAME

```

for historic slurm job logs:

```
#!/bin/env bash

JOBNAME=mytarget_0888390

sacct -l --name $JOBNAME

```

Should be fairly easy to run as a system call within targets using static branching.

**Reply by multimeric** [reply]
> Any R functions for log management would need to be available through future, batchtools, or future.batchtools because those are the packages that are closer to the level at which those logs are generated.

Right, but `batchtools` *does* provide an API for accessing the logs, via the [`getLog` function](https://mllg.github.io/batchtools/reference/showLog.html). But it sounds like you're hesitant to integrate anything that doesn't use the futures interface. In the short term I can probably write a function that takes a targets task name, looks up the job ID and calls `getLog` on it, but it would be nice for this to be built into `targets`.

**Reply by wlandau** [reply]
It's out of scope for `targets`.

`future.batchtools` would be a more appropriate place for a function like this, or ‘batchtools’ itself. And then you would call those packages directly instead of going through `targets`. You would use known job names, which in this case are the target names.
